### ISSUE TITLE
What is the difference between p2,p3,p4 output?

### ISSUE THREAD

**Issue by billamiable** [chuhang/SurfConv#1]
Hi, 

I like your work and here is a naive question:
I found the code output of "resample_input.py" is to duplicate single input image into 4-level images. I noticed that p1 is smaller in scale compared to the latter three. However, I didn't see much difference between p2, p3, p4. Could you please explain that?

Thanks!

**Comment by chuhang** [comment]
Hi, thank you so much for your interest in our work.

p1-p4 are supposed to be the same input image at different scale, so p2, p3, p4 have different resolutions. For the level-wise mask part, we put it into the label image for simplicity.

Thank you for the question. We'll update the code comments to make it clearer.

**Comment by billamiable** [comment]
Get it, one more naive question plz:
During training, will we just concatenate p1-p4 as input? And meanwhile, just according to visual effect, I observe that the resolution difference between p2-p4 is relatively small, i wonder why this transformation will enhance performance? (I didn't understand the paper thoroughly, so it might be a very simple misunderstanding)
Thanks!

**Comment by chuhang** [comment]
Thanks for the following-up. During training, we just back propagate the same convnet through all levels (4 as in example, or more) and update parameters.

The benefit of this mainly comes from scale invariance. Think of a car nearby and another car far-away. By this depth-aware operation, they are (approximately) put into the same metric scale. So the network doesn't need to learn two patterns in its receptive field of the car (as in standard 2d-conv), but instead one (as in ours).

**Comment by billamiable** [comment]
Ok, Thanks for your prompt reply!

**Comment by billamiable** [comment]
Hi, one more question about implementation~

I wanna try this code in my own network, is the following pipeline right?
1. use D4 to generate different scale images;
2. set up N (for example, 4) cnn network to process different scale inputs and add their loss together to form the final training loss (as shown in image below).
![webwxgetmsgimg](https://user-images.githubusercontent.com/16483494/41590017-9012b00c-73e8-11e8-9f38-6ef748f270fd.jpeg)

am i right? Thanks a lot!

**Comment by chuhang** [comment]
Hi really glad that you are giving it a try. You are almost there, only one difference: use one same cnn (with shared weights) instead of using 4 cnns. This way only one set of weights is used across all depth levels. Also in D4 there is a gamma value which is quite important to tune and dataset sensitive (see paper for detail)

If there’s plenty of data to ensure sufficient training in your case, 4 cnns may actually give even better results. We haven’t tried yet.

**Comment by billamiable** [comment]
Oh, i see! Thanks for your kind reminder. 

So does that mean i will train the four-scale inputs one-by-one in an iteration? If so, the loss generated by each depth level will no longer need to be summed, right? Because only 4 cnns can enable simultaneously training four-scale inputs.

**Comment by chuhang** [comment]
Hi billamiable, I was travelling last week, sorry for the delay.

In our code, we train only one CNN simultaneously on four-scale inputs. Please refer to our training script for the details. Basically, we forward the CNN on 4 different scales, accumulating the gradient, and backward altogether.

Training 4 CNNs individually for each scale (partitioned with our D4) is out of the scope of our original SurfConv, as it will not have the same model parameter efficiency (4x weights in final model). However, if you have large amount of data and don't care much about model size, please feel free to explore anything.

**Comment by billamiable** [comment]
That's okay, my friend!
Ok, thanks a lot for your kind reply~ I will try to figure out and see if i have more questions :)
